backend/inspirehep/search/factories/query.py

```diff
@@ -4,37 +4,58 @@
 #
 # inspirehep is free software; you can redistribute it and/or modify it under
 # the terms of the MIT License; see LICENSE file for more details.
-import re
 
 import inspire_query_parser
 from elasticsearch_dsl import Q
 from flask import current_app
 
-from inspirehep.pidstore.api.base import PidStoreBase
+from inspirehep.pidstore.api import PidStoreBase
 from inspirehep.search.utils import RecursionLimit
 
 
-def _convert_citedby_query(query_string):
-    citeby_matches = re.findall(
-        r"(?<=citedby:recid:)\d{1,}", query_string, re.IGNORECASE
-    )
-    if not citeby_matches:
-        return
-    for recid in citeby_matches:
-        record_uuid = PidStoreBase.get_uuid_for_recid(recid, "lit")
-        if record_uuid:
-            query_string = re.sub(
-                f"(?<=citedby:recid:){recid}", str(record_uuid), query_string
-            )
-    return query_string
+def replace_recid_in_citedby_query(query):
+
+    citedby_query_fields = set(["index", "id", "path"])
+    self_ref_field = "self.$ref.raw"
+
+    def _replace_recid_in_citedby_query(obj):
+        if isinstance(obj, dict):
+            for key, val in obj.items():
+                if isinstance(val, list):
+                    _replace_recid_in_citedby_query(val)
+                elif isinstance(val, dict):
+                    # check if we are in citedby query and replace
+                    # recid with uuid
+                    if (
+                        citedby_query_fields == set(val.keys())
+                        and self_ref_field in obj.keys()
```

Comment:
I would move the obj.keys() outside, since it's always the same and we don't have to calculate it in every iteration